src/python/TaskWorker/Actions/DBSDataDiscovery.py

```diff
@@ -142,14 +165,60 @@ def executeInternal(self, *args, **kwargs):
         ## '/JetHT/Run2016B-PromptReco-v2/AOD#89b03ca6-1dc9-11e6-b567-001e67ac06a0': [u'T1_IT_CNAF_Buffer', u'T2_US_Wisconsin', u'T1_IT_CNAF_MSS', u'T2_BE_UCL']}
         try:
             dbsOnly = self.dbsInstance.split('/')[1] != 'global'
-            locationsMap = self.dbs.listFileBlockLocation(list(blocks), dbsOnly=dbsOnly)
-            if secondaryDataset:
-                secondaryLocationsMap = self.dbs.listFileBlockLocation(list(secondaryBlocks), dbsOnly=dbsOnly)
+            # TODO:
+            #scope = "cms" if dbsOnly else "user.%s" % username 
+            scope = "cms"
+            self.logger.info("Trying data location with Rucio first")
+
+            locations = rucioClient.getReplicaInfoForBlocks(scope=scope, block=list(blocks))
+            located_blocks = locations['phedex']['block']
+            locationsMap = {}
+
+            if len(located_blocks) == 0:
+                raise Exception('No location found for blocks')
+
+            for element in located_blocks:
+                locationsMap.update({element['name']: [ x['node'] for x in element['replica'] ] })
+
+            #self.logger.info("LOCATIONS RUCIO: %s" % locationsMap)
         except Exception as ex: # TODO should we catch HttpException instead?
-            self.logger.exception(ex)
-            raise TaskWorkerException("The CRAB3 server backend could not get the location of the files from dbs or phedex.\n"+\
-                                      "This is could be a temporary phedex/dbs glitch, please try to submit a new task (resubmit will not work)"+\
-                                      " and contact the experts if the error persists.\nError reason: %s" % str(ex))
+            self.logger.warn("No locations foud with rucio: %s \n Trying with DBS and PhEDEx" % str(ex))
+            try:
+                locationsMap = self.dbs.listFileBlockLocation(list(blocks), dbsOnly=dbsOnly)
+            except Exception as ex:
+                raise TaskWorkerException(
+                    "The CRAB3 server backend could not get the location of the files from dbs or phedex or rucio.\n"+\
+                    "This is could be a temporary phedex/rucio/dbs glitch, please try to submit a new task (resubmit will not work)"+\
+                    " and contact the experts if the error persists.\nError reason: %s" % str(ex)
+                    )
+        if secondaryDataset:
+            try:
+                dbsOnly = self.dbsInstance.split('/')[1] != 'global'
+                # TODO:
+                #scope = "cms" if dbsOnly else "user.%s" % username 
+                scope = "cms"
```

Comment:
Good point. Now... forgive me if I am simply getting lost. Does one dataset exist in one scope only ? If I as a user set a Rucio rule to keep a replica of some nanoAOD in my user space and quota, does it mean that the dataset is both scopes ? And is there the possibilit that the dataset only exist in my scope at some point ?

Reply:
@ericvaandering , yes, I put the TODO in the comments for this reason, but since we are limiting data location to NanoAOD* only, this could be a separate issue I think.

@belforte AFAICT (but I could be wrong) the dataset scope and the rules are 2 different entities in RUCIO. You can have everything you mentioned before. The scope for the dataset is set at registration time, where quotas and permissions are checked, so indeed a user produced dataset will always be on one user scope only, while CMS produced one will always be in cms scope.
Said that, putting rules on cms scope will also be accounted as you said, but I cannot tell the dynamic there.

Reply:
Right, Stefano, you are mixing up scope which is a property of dataset and quota which is a property of (RSE, user) pair.

So a dataset is created with a scope and that's immutable. If you set a rule to keep the dataset at your favorite RSE, you use quota at that RSE, but it's the same dataset stored in the same LFNs and PFNs. And if I ask for a copy at the same RSE, there is still only one copy of each file.

Reply:
thanks both for clarification.

Reply:
@dciangot it may be good to put code in to select user scope when appropriate already, even if we do not use immediately, simply because you now have everything fresh in mind. I do not thinkg that "dbsOnly" is the appropriate switch, even if it worked accidentally.

Reply:
mmm, can you expand a bit? If understand correctly, dbsOnly==False flag here means that the dataset is not on DBS Global instance that should be in principle overlapping with datasets in `cms` scope.

While, if dbsOnly==True the dataset is a user  published one, so should land on a `user.<username>` scope. Here, we have the problem that we might need the user to indicate what username to look at, since it can be that he wants to read a dataset under another user scope. 
Since this is not needed at this time (we only want data location for cms NanoAOD), I created a separate Issue , and we should be ok with keeping CMS only in the scope for now.

What do you think, @belforte @ericvaandering ?

Reply:
yes, it works. But imagining a world w/o DBS, a variable named `dbsOnly` makes no sense. While Rucio will still be there. Dont't you get same resutl with (pseudocode)
 `if datasetname.endswith('USER')`
?
anyhow, sure. Leave it with CMS scope only and let's follow up in the new issue